File: larcenous/week_11/BOJ_26152_플래피 버드 스코어링.py
N = int(input())
A = list(map(int,input().split()))
B = list(map(int,input().split()))
S = []
Q = int(input())
W = list(map(int,input().split()))

for i,j in zip(A,B):
    S.append(i-j)
for i in range(1, len(S)): #단조 감소 처리를 통해 내림차순으로 만듦, 대신 중복된 원소가 존재하므로 upper bound를 찾는 이분탐색이 됨
        if S[i] > S[i-1]:
            S[i] = S[i-1]
S = [-i for i in S] #내림차순 정렬을 오름차순 정렬로 변경

def binary_search(left, right, target):
    while left < right:
        mid = (left + right) // 2
        if S[mid] <= target:
            left = mid + 1
        else:
            right = mid
    return left

for bird in W :
    print(binary_search(0,len(S),-bird))

File: slavicmeme/week_11/BOJ_26152_플래피버드스코어링.java
package Week11;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BOJ_26152_플래피버드스코어링 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int[] A = new int[N];
        int[] B = new int[N];
        int[] C = new int[N];
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++)
            A[i] = Integer.parseInt(st.nextToken());
        
        st = new StringTokenizer(br.readLine());
        
        for (int i = 0; i < N; i++) 
        {
            B[i] = Integer.parseInt(st.nextToken());
            C[i] = A[i] - B[i];
        }
        
        for (int i = 1; i < N; i++)
            if (C[i] > C[i - 1])
                C[i] = C[i - 1];
        
        for (int i = 0; i < N; i++)
            C[i] = -C[i];
        
        Arrays.sort(C);

        int Q = Integer.parseInt(br.readLine());
        int[] W = new int[Q];
        
        st = new StringTokenizer(br.readLine());
        
        for (int i = 0; i < Q; i++)
            W[i] = Integer.parseInt(st.nextToken());

        for (int bird : W) 
        {
            int result = binarySearch(C, -bird);
            System.out.println(result);
        }

        br.close();
    }

    private static int binarySearch(int[] arr, int target) 
    {
        int left = 0;
        int right = arr.length;
        while (left < right) 
        {
            int mid = (left + right) / 2;
            if (arr[mid] <= target)	
            	left = mid + 1;
            else	
            	right = mid;
        }
        return left;
    }
}
